mlir/lib/Transforms/RemoveDeadValues.cpp

```diff
@@ -754,17 +760,50 @@ static void cleanUpDeadVals(RDVFinalCleanupList &list) {
     // Some functions may not allow erasing arguments or results. These calls
     // return failure in such cases without modifying the function, so it's okay
     // to proceed.
-    (void)f.funcOp.eraseArguments(f.nonLiveArgs);
+    if (succeeded(f.funcOp.eraseArguments(f.nonLiveArgs))) {
+      // Record only if we actually erased something.
+      if (f.nonLiveArgs.any())
+        erasedFuncArgs.try_emplace(f.funcOp.getOperation(), f.nonLiveArgs);
+    }
     (void)f.funcOp.eraseResults(f.nonLiveRets);
   }
 
   // 4. Operands
   LDBG() << "Cleaning up " << list.operands.size() << " operand lists";
   for (OperationToCleanup &o : list.operands) {
-    if (o.op->getNumOperands() > 0) {
-      LDBG() << "Erasing " << o.nonLive.count()
-             << " non-live operands from operation: "
-             << OpWithFlags(o.op, OpPrintingFlags().skipRegions());
+    if (auto call = dyn_cast<CallOpInterface>(o.op)) {
+      // Use the stored callee reference if available, avoiding expensive symbol
+      // lookup
+      if (o.callee) {
```

Comment:
```suggestion
      if (o.callee) {
        // Use the stored callee reference if available, avoiding expensive symbol
        // lookup
        auto call = cast<CallOpInterface>(o.op);
```

This embeds an assertion that o.callee is only set for calls.

Reply:
True, added an explicit check to only apply the call related logic if it is one.